File: src/etf_modules/sim_bringup/scripts/test_move_xarm6.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "trajectory_msgs/msg/joint_trajectory.hpp"
#include "trajectory_msgs/msg/joint_trajectory_point.hpp"

using namespace std::chrono_literals;

class TestPlannersNode : public rclcpp::Node
{
private:
    rclcpp::TimerBase::SharedPtr timer;
    rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr trajectory_publisher;

public:
    TestPlannersNode() : Node("trajectory_publisher_node")
    {
        trajectory_publisher = this->create_publisher<trajectory_msgs::msg::JointTrajectory>("/xarm6_traj_controller/joint_trajectory", 10);
        timer = this->create_wall_timer(15s, std::bind(&TestPlannersNode::test_planners_callback, this));
    }

private:
    void test_planners_callback()
    {
        trajectory_msgs::msg::JointTrajectory trajectory;
        trajectory.joint_names = {"joint1", "joint2", "joint3", "joint4", "joint5", "joint6"};
        trajectory_msgs::msg::JointTrajectoryPoint point;

        point.positions = {0, 0, 0, 0, 0, 0};
        point.time_from_start.sec = 0;
        trajectory.points.emplace_back(point);
        
        point.positions = {M_PI_2, 0, 0, 0, 0, 0};
        point.time_from_start.sec = 5;
        trajectory.points.emplace_back(point);

        point.positions = {0, 0, 0, 0, 0, 0};
        point.time_from_start.sec = 10;
        trajectory.points.emplace_back(point);

        std::cout << "Publishing trajectory ...\n";
        trajectory_publisher->publish(trajectory);
    }
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TestPlannersNode>());
    rclcpp::shutdown();
    return 0;
}

File: src/etf_modules/sim_bringup/src/test_move_xarm6.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import rclpy
from rclpy.node import Node
from builtin_interfaces.msg import Duration

from std_msgs.msg import String
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint


class TestJointTrajectoryNode(Node):

    def __init__(self):
        super().__init__('trajectory_publisher_node')
        publish_topic = '/xarm6_traj_controller/joint_trajectory'
        self.trajectory_publihser = \
            self.create_publisher(JointTrajectory, publish_topic, 10)
        timer_period = 5.0
        self.timer = self.create_timer(timer_period,
                self.timer_callback)
        self.joints = [
            'joint1',
            'joint2',
            'joint3',
            'joint4',
            'joint5',
            'joint6',
            ]

    def timer_callback(self):
        trajectory = JointTrajectory()
        trajectory.joint_names = self.joints
        joint_point1 = JointTrajectoryPoint()
        joint_point1.positions = [
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            0.0,
            ]
        joint_point1.time_from_start = Duration(sec=1)
        trajectory.points.append(joint_point1)

        joint_point2 = JointTrajectoryPoint()
        joint_point2.positions = [
            0.5,
            0.0,
            -0.6,
            0.0,
            0.0,
            0.0,
            ]
        joint_point2.time_from_start = Duration(sec=2)
        trajectory.points.append(joint_point2)

        joint_point3 = JointTrajectoryPoint()
        joint_point3.positions = [
            -0.5,
            0.0,
            -0.6,
            0.0,
            0.0,
            0.0,
            ]
        joint_point3.time_from_start = Duration(sec=3)
        trajectory.points.append(joint_point3)

        self.trajectory_publihser.publish(trajectory)


def main(args=None):
    rclpy.init(args=args)
    joint_publisher = TestJointTrajectoryNode()
    rclpy.spin(joint_publisher)
    joint_publisher.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
